### PR TITLE
Users/padepu/azure rm linux web app support

### PR DESCRIPTION
**Pull request by pavanadepu2**
Added the implementation for supporting container based web deployments for Linux.

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 1]
```diff
@@ -16,25 +16,35 @@
     "version": {
         "Major": 3,
         "Minor": 2,
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: Can you bump up minor version. I believe the new Linux support is not just a patch update.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 2]
```diff
@@ -16,25 +16,35 @@
...
-        "Patch": 2
+        "Patch": 3
     },
+    "preview": "true",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: We have removed preview tag from this task recently. Don't reintroduce it.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 3]
```diff
@@ -58,12 +68,77 @@
...
+            "required": false,
+            "visibleRule": "WebAppName = invalidwebappname",
+            "properties": {
+                "EditableOptions": "false"
+            }
+        },   
+        {
+            "name": "DockerNamespace",
+            "type": "string",
+            "label": "Registry\\Namespace",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: Would it be better Registry (or) NameSpace. Check with PMs.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 4]
```diff
@@ -58,12 +68,77 @@
...
+            "helpMarkDown": "Docker repository namespace."
+        },
+        {
+            "name": "DockerRepository",
+            "type": "string",
+            "label": "Repository",
+            "defaultValue": "",
+            "required": true,
+            "properties": {
+                "EditableOptions": "false"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: Why is this text input non-editable?

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 5]
```diff
@@ -58,12 +68,77 @@
...
+            "helpMarkDown": "Docker repository or image name."
+        },
+        {
+            "name": "DockerImageTag",
+            "type": "string",
+            "label": "Tag",
+            "defaultValue": "",
+            "required": false,
+            "properties": {
+                "EditableOptions": "false"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: Why is tag name non-editable?

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 6]
```diff
@@ -171,6 +248,18 @@
...
         {
+            "name": "AppSettings",
+            "type": "multiLine",
+            "label": "App settings",
+            "defaultValue": "",
+            "required": false,
+            "groupName": "ApplicationSettings",
+            "helpMarkDown": "Webapp application settings.",
+            "properties": {
+                "editorExtension": "ms.vss-services-azure.parameters-grid"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: Can you check with @ds-ms . He was planning to convert these extensions to react.

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 7]
```diff
@@ -60,107 +62,121 @@ async function run() {
             resourceGroupName = await azureRESTUtility.getResourceGroupName(endPoint, webAppName);
         }
 
-        var publishingProfile = await azureRESTUtility.getAzureRMWebAppPublishProfile(endPoint, webAppName, resourceGroupName, deployToSlotFlag, slotName);
-        console.log(tl.loc('GotconnectiondetailsforazureRMWebApp0', webAppName));
+        // For container based linux deployment
+        if(webAppKind && webAppKind === "linux")
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: This condition will break the ability to do in-built image deployment to Linux web apps. Can you check condition that is specific to container based deployment?

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 8]
```diff
@@ -60,107 +62,121 @@ async function run() {
...
+        {
+            tl.debug("Performing the deployment of webapp for Linux.");
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: "Performing container based deployment...."

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    if(imageSourceAndTag)
+    {
+        tl.debug("Deploying the image " + imageSourceAndTag + " to the webapp " + webAppName);
+        
+        appSettings = appSettings ? appSettings.trim() : "";        
+        appSettings = "-DOCKER_CUSTOM_IMAGE_NAME " + imageSourceAndTag + " " + appSettings;
+    
+        // Update webapp application setting
+        var webAppSettings = await azureRESTUtility.getWebAppAppSettings(endPoint, webAppName, resourceGroupName, false, null);
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Don't you need to consider slot?

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        imageSourceAndTag = dockerNamespace + "/" + dockerRepository + ":" + dockerImageTag;          
+    } else {
+        imageSourceAndTag = dockerNamespace + "/" + dockerRepository;
+    }
+
+    if(imageSourceAndTag)
+    {
+        tl.debug("Deploying the image " + imageSourceAndTag + " to the webapp " + webAppName);
+        
+        appSettings = appSettings ? appSettings.trim() : "";
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Can user provide DOCKER_CUSTOM_IMAGE_NAME in appSettings via Task input? Should we error out if present?

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    if(imageSourceAndTag)
+    {
+        tl.debug("Deploying the image " + imageSourceAndTag + " to the webapp " + webAppName);
+        
+        appSettings = appSettings ? appSettings.trim() : "";        
+        appSettings = "-DOCKER_CUSTOM_IMAGE_NAME " + imageSourceAndTag + " " + appSettings;
+    
+        // Update webapp application setting
+        var webAppSettings = await azureRESTUtility.getWebAppAppSettings(endPoint, webAppName, resourceGroupName, false, null);
+        mergeAppSettings(appSettings, webAppSettings);
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: You are not assigning the return value from this function.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        {            
+            tl.debug("Updating the startup command: " + startupCommand);
+            var updatedConfigDetails = JSON.stringify(
+            {
+                "properties": {
+                    "appCommandLine": startupCommand
+                }
+            });
+
+            await azureRESTUtility.updateAzureRMWebAppConfigDetails(endPoint, webAppName, resourceGroupName, false, null, updatedConfigDetails);
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Same, is slot not interesting?

File: Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson [hunk 13]
```diff
@@ -6,10 +6,21 @@
   "loc.group.displayName.FileTransformsAndVariableSubstitution": "File Transforms & Variable Substitution Options",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson]: Can you add L0 tests for the changes done.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 12 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Currently Slot is not supported for Linux deployments.

File: Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson (hunk 13 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson]: Will be sending a separate PR for it.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 11 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Actually return doesn't make sense as the WebAppSettings itself got updated. Removed the return from the mergeAppSettings function.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 9 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Currently Slot is not supported for Linux deployments.

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts (hunk 8 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Done

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts (hunk 7 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Considering the dockerNamespace.

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 1 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/task.json]: Missed in commit, updated.

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 2 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/task.json]: Removed.

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 4 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/task.json]: string type has impact of properties. Forgot to remove while changing from pickList to string.

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 5 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/task.json]: explained above.

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 6 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/task.json]: @ds-ms is taking care of back compact.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 9 above)
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Instead of hard coding the slot param to false, should you then pass the flag we get for slot input as is. Right now it will be false anyway, later if it is supported, you get the benefit for free.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 12 above)
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: Same comment as above.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 10 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: AppSettings will always take the higher precedence here.

The namespace may become non editable in the next version. This feature allows the users to specify their own public image name when the Namespace field becomes non-editable.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 9 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: The value of slot will be persisted  if user selects a usual webapp before selecting the linux one. So, to handle such situations, I am sending false. Will update it with slotId, when we support slots.

File: Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts (hunk 12 above)
Comment (pavanadepu2) [inline Tasks/AzureRmWebAppDeployment/azurermwebappcontainerdeployment.ts]: explained above.

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 14]
```diff
@@ -63,65 +66,76 @@ async function run() {
         var publishingProfile = await azureRESTUtility.getAzureRMWebAppPublishProfile(endPoint, webAppName, resourceGroupName, deployToSlotFlag, slotName);
```
Comment (asranja) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Make the deployToSlotFlag false in case of linux apps. Caching issues.

Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Done

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 15]
```diff
@@ -15,28 +15,39 @@
...
             "displayName": "Output",
-            "isExpanded": true
+            "isExpanded": true,
+            "visibleRule": "WebAppKind = app || WebAppKind = functionapp"
+        },
+        {
+            "name": "ApplicationSettings",
+            "displayName": "Application Settings",
+            "isExpanded": true,
+            "visibleRule": "WebAppKind = linux"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: This change is fine for now. Later when we support built-in linux deployment, this option's visibility should be controlled.

Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/task.json]: Sure, will do it while supporting built-in Linux deployment.
